### PR TITLE
HIVE-24855: Introduce virtual colum ROW__IS__DELETED

### PR DESCRIPTION
**Pull request by kasakrisz**
### What changes were proposed in this pull request?
1. Add new Virtual column `ROW__IS__DELETED`
2. Add new setting `hive.acid.fetch.deleted.rows` to pass property to `OrcInputFormat` whether deleted rows should be fetched
3. Refactor adding virtual columns in LlapInputFormat: able to handle multiple virtual columns and the list can be appended easier
4. Add new hint `FETCH_DELETED_ROWS(<comma separated table list>)` to be able to enable this feature from query text.
5. Add property `fetchDeletedRowsScans` to `ql/Context.java` to store which TableScans should fetch deleted rows.
6. Change keytype from `RecordIdentifier` to `OrcRawRecordMerger.ReaderKey` because latter contains an isDeleteEvent property.

### Why are the changes needed?
This patch is part of HIVE-24854: Incremental Materialized view refresh in presence of update/delete operations

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
```
mvn test -Dtest.output.overwrite -DskipSparkTests -Dtest=TestMiniLlapLocalCliDriver -Dqfile=fetch_deleted_rows.q -pl itests/qtest -Pitests
```

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/io/BatchToRowReader.java [hunk 1]
```diff
@@ -108,25 +108,41 @@ public BatchToRowReader(RecordReader<NullWritable, VectorizedRowBatch> vrbReader
...
+      }
     }
+
     if (LOG.isDebugEnabled()) {
       LOG.debug("Including the columns " + DebugUtils.toString(included));
     }
     this.included = included;
   }
 
+  public static class VirtualColumnHandler {
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/io/BatchToRowReader.java:127]: Can you add a short comment for the inner class?

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 2]
```diff
@@ -4450,6 +4450,8 @@ private static void populateLlapDaemonVarsSet(Set<String> llapDaemonVarsSetLocal
     HIVE_ACID_DIRECT_INSERT_ENABLED("hive.acid.direct.insert.enabled", true,
         "Enable writing the data files directly to the table's final destination instead of the staging directory."
         + "This optimization only applies on INSERT operations on ACID tables."),
+    HIVE_ACID_FETCH_DELETED_ROWS("hive.acid.fetch.deleted.rows", false,
```
[the remark this answers is not in the document]
Comment (kasakrisz) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java]: moved to constants

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 3]
```diff
@@ -1970,8 +1970,22 @@ private int processQueryHint(ASTNode ast, QBParseInfo qbp, int posn) throws Sema
     String queryHintStr = ast.getText();
     LOG.debug("QUERY HINT: {} ", queryHintStr);
     try {
-      ASTNode hintNode = pd.parseHint(queryHintStr);
-      qbp.setHints(hintNode);
+      ASTNode hintListNode = pd.parseHint(queryHintStr);
+      qbp.setHints(hintListNode);
+      for (int i = 0; i < hintListNode.getChildCount(); ++i) {
+        ASTNode hintNode = (ASTNode) hintListNode.getChild(i);
+        if (hintNode.getChild(0).getType() != HintParser.TOK_FETCH_DELETED_ROWS) {
```
[the remark this answers is not in the document]
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: Removed hint. Using `acid.fetch.deleted.rows` as table property.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 4]
```diff
@@ -1943,15 +1943,17 @@ public static boolean isFullAcidScan(Configuration conf) {
    *                   we assume this is a full transactional table.
    */
   public static void setAcidOperationalProperties(
-      Configuration conf, boolean isTxnTable, AcidOperationalProperties properties) {
+      Configuration conf, boolean isTxnTable, AcidOperationalProperties properties, boolean fetchDeletedRows) {
```
[the remark this answers is not in the document]
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1952]: Added back the old signature.

File: ql/src/java/org/apache/hadoop/hive/ql/io/BatchToRowReader.java (hunk 1 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/io/BatchToRowReader.java:132]: Added.

File: llap-server/src/java/org/apache/hadoop/hive/llap/io/api/impl/LlapInputFormat.java [hunk 5]
```diff
@@ -182,11 +190,11 @@ static VectorizedRowBatchCtx createFakeVrbCtx(MapWork mapWork) throws HiveExcept
...
-    boolean hasRowId = false;
+    ArrayList<VirtualColumn> virtualColumnList = new ArrayList<>(2);
     for (ColumnInfo c : rowSchema.getSignature()) {
       String columnName = c.getInternalName();
-      if (VirtualColumn.ROWID.getName().equals(columnName)) {
-        hasRowId = true;
+      if (ALLOWED_VIRTUAL_COLUMNS.containsKey(columnName)) {
+        virtualColumnList.add(ALLOWED_VIRTUAL_COLUMNS.get(columnName));
       } else {
         if (VirtualColumn.VIRTUAL_COLUMN_NAMES.contains(columnName)) continue;
```
[the remark this answers is not in the document]
Comment (kasakrisz) [reply llap-server/src/java/org/apache/hadoop/hive/llap/io/api/impl/LlapInputFormat.java]: merged

File: ql/src/test/queries/clientpositive/fetch_deleted_rows.q [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+create table t1(a int, b varchar(128)) stored as orc tblproperties ('transactional'='true');
+
+insert into t1(a,b) values (1, 'one'), (2, 'two');
+
+delete from t1 where a = 1;
+
+insert into t1(a,b) values (3, 'three'), (4, 'four'), (4, 'four again'), (5, 'five');
+
+alter table t1 set tblproperties ('acid.fetch.deleted.rows'='true');
+select t1.ROW__IS__DELETED, * from t1;
```
Comment (jcamachor) [inline ql/src/test/queries/clientpositive/fetch_deleted_rows.q:14]: Just leaving a note. I think this syntax passing the table properties in the query should also work (if that's the case, maybe we can add a query with it):
```
select t1.ROW__IS__DELETED, *
from t1('acid.fetch.deleted.rows'='true');
```

**Review by jcamachor — APPROVED** [review]
Left a minor comment/addition. Other than that, LGTM, +1

File: ql/src/test/queries/clientpositive/fetch_deleted_rows.q (hunk 6 above)
Comment (kasakrisz) [reply ql/src/test/queries/clientpositive/fetch_deleted_rows.q:14]: When the logical plan is built from the AST and `HiveTableScans` are instantiated table properties defined in the AST are lost.
When the optimized logical plan is converted back to AST only properties coming from Druid and JDBC scans are populated. We could add additional properties from `HiveTableScans` in the same way.

I think passing `acid.fetch.deleted.rows` as a table property in the query is a good idea and we can benefit from it  when building an Incremental MV rebuild plan. I will try to include this into [HIVE-24854](https://issues.apache.org/jira/browse/HIVE-24854)

**Pull request merged**